operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1791 @@
+	go func() {
+		ipam.wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+	case <-ctx.Done():
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) restart() {
+	ipam.logger.Info("Restarting LB IPAM")
+
+	// Refresh the context
+	ipam.ctx, ipam.cancel = context.WithCancel(context.Background())
+
+	// Reset all stored state
+	ipam.pools = make(map[string]*cilium_api_v2alpha1.CiliumLoadBalancerIPPool)
+	ipam.rangesStore = newRangesStore()
+	ipam.serviceStore = NewServiceStore()
+
+	// Re-start the main goroutine
+	ipam.wg.Add(1)
+	go func() {
+		defer ipam.wg.Done()
+		ipam.Run()
+	}()
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*slim_core_v1.Service]
+
+func (ipam *LBIPAM) Run() {
+	// PreInit blocks and waits until there is at least 1 pool, which is the trigger to activate the controller.
+	ipam.logger.Info("LB-IPAM in pre-init phase, waiting for pools")
+	err := ipam.preInit()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while pre-initializing LB-IPAM")
+		return
+	}
+	// If we exited preInit because we are shutting down, just return
+	if ipam.ctx.Err() != nil {
+		return
+	}
+
+	errChan := make(chan error, 2)
+	poolChan := stream.ToChannel[ipPoolEvent](ipam.ctx, errChan, ipam.poolResource)
+	svcChan := stream.ToChannel[svcEvent](ipam.ctx, errChan, ipam.svcResource)
+
+	ipam.logger.Info("LB-IPAM initializing")
+	err = ipam.init(poolChan, svcChan)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while initializing LB-IPAM")
+		// Stop, don't go into main control loop
+		return
+	}
+
+	ipam.logger.Info("LB-IPAM done initializing")
+	for _, cb := range ipam.initDoneCallbacks {
+		if cb != nil {
+			cb()
+		}
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are fully drained to avoid
+	// blocking the resource and leaking goroutines during testing.
+	for poolChan != nil || svcChan != nil {
+		select {
+		case event, ok := <-poolChan:
+			if !ok {
+				poolChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.poolOnUpsert,
+				ipam.poolOnDelete,
+			)
+
+			// This controller must go back into a dormant state when the last pool has been removed
+			if len(ipam.pools) == 0 {
+				// Trigger the shutdown of this routine
+				ipam.cancel()
+				// Upon return, restart the controller, which will start in pre-init state
+				defer ipam.restart()
+			}
+
+		case event, ok := <-svcChan:
+			if !ok {
+				svcChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.svcOnUpsert,
+				ipam.svcOnDelete,
+			)
+		case err := <-errChan:
+			if err == nil {
+				continue
+			}
+
+			ipam.logger.WithError(err).Error("Unrecoverable error, triggering shutdown")
+			ipam.shutdowner.Shutdown(hive.ShutdownWithError(err))
+		}
+	}
+}
+
+// preInit blocks until a pool object exists in the store
+func (ipam *LBIPAM) preInit() error {
+	preInitCtx, cancel := context.WithCancel(ipam.ctx)
```

Comment:
hmm, the ctx usage here is odd. 

Really, the constructor of LB-IPAM should provide a context to the "Start" method, and then it should be passed thru the call stack. 

For instance, preInit should receive a "ctx" and if then if it needs some timeout behavior internally, it can derive a context from it. 

We probably should not be holding onto `ipam.ctx` and instead thread a context thru from the Start() method and on.

Reply:
This ctx is here to manage the lifetime of `stream.ToChannel` so we can signal it to stop. It is a sub-ctx of `ipam.ctx` to that it is canceled when the outer ctx is canceled.

We do not want to timeout here, we want to intentionally block and not start the full control until we see at least 1 Pool which allows us to always enable the feature (no flag) without much overhead.

When we do see a pool, we cancel this `ToChannel` and continue on the the actual init process. I believe this is a very valid usage of a context.

Reply:
The context usage is fine here with `ToChannel`. I would however consider removing the whole `preInit` and just subscribing to pools only once (e.g. in Run() line 214) and then waiting for the first to show up before subscribing to services. You might have little bit of duplication with handling the first event, but it should still be less code overall.

Reply:
You could even do this from within the `for select` loop and swap `svcChan` into a real one from an empty one.